alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs

```diff
@@ -413,3 +420,39 @@ newtype PPHash crypto
 deriving newtype instance CC.Crypto crypto => FromCBOR (PPHash crypto)
 
 deriving newtype instance CC.Crypto crypto => ToCBOR (PPHash crypto)
+
+hashLanguagePP :: forall era. Era era => PParams era -> Language -> PPHash (Crypto era)
+hashLanguagePP pp lang = PPHash (hash2P (Proxy @EraIndependentPP) (Proxy @era) [HashItem restrictedmap])
+  where
+    restrictedmap :: Map.Map Language CostModel
+    restrictedmap = eval (singleton lang ◁ (_costmdls pp))
+
+-- ===============================================================
+-- Hashing Tuples (upto to width 4) in a uniform way
+
+data HashItem where
+  HashItem :: ToCBOR a => a -> HashItem
+
+serializeHashTuple :: [HashItem] -> ByteString
+serializeHashTuple xs = mconcat (map f xs)
+  where
+    f (HashItem x) = serialize' x
+
+-- Hash with 2 Proxys. This function is only used internally
+-- The type families need alot of help keepng things straight,
+-- the Proxys (and the type signature) provide that help.
+
+hash2P :: forall t e. Era e => Proxy t -> Proxy e -> [HashItem] -> Hash.Hash (CC.HASH (Crypto e)) t
+hash2P _ _ x = Hash.castHash (Hash.hashWith serializeHashTuple x)
+
+class Era e => Hashable e t where
```

Comment:
we are using the `Hashable` typeclass from `Data.Hashable` (hashable package) elsewhere in our codebase, so we will need a new name.

Reply:
I changed it to HashTuple, since all instances are tuples.